Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs

```diff
@@ -0,0 +1,232 @@
+        }
+
+        [Test]
+        public void Should_ReturnFalse_InWaitForExist_WhenElementDoesNotExist()
+        {
+            Assert.IsFalse(notPresentLabel.State.WaitForExist(TimeSpan.Zero));
+        }
+
+        [Test]
+        public void Should_ReturnTrue_InWaitForExist_WhenElementExists()
+        {
+            Assert.IsTrue(RightArgumentTextBox.State.WaitForExist(TimeSpan.Zero));
+        }
+
+        [Test]
+        public void Should_ReturnFalse_InWaitForNotExist_WhenElementExists()
+        {
+            Assert.IsFalse(RightArgumentTextBox.State.WaitForNotExist(TimeSpan.Zero));
+        }
+
+        [Test]
+        public void Should_ReturnTrue_InWaitForNotExist_WhenElementDoesNotExist()
+        {
+            Assert.IsTrue(notPresentLabel.State.WaitForNotExist(TimeSpan.Zero));
+        }
+
+        [Test]
+        public void Should_WaitForClickable_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            try
+            {
+                EmptyButton.State.WaitForClickable(FromSeconds);
+            }
+            catch (WebDriverTimeoutException)
+            {
+                AssertIfWaitTimeIsCorrect(stopwatch);
+            }
+        }
+
+        [Test]
+        public void Should_WaitForDisplayed_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            MaximizeButton.State.WaitForDisplayed(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        [Test]
+        public void Should_WaitForEnabled_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            EmptyButton.State.WaitForEnabled(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        [Test]
+        public void Should_WaitForExist_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            notPresentLabel.State.WaitForExist(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        [Test]
+        public void Should_WaitForNotExist_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            RightArgumentTextBox.State.WaitForNotExist(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        [Test]
+        public void Should_WaitForNotDisplayed_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            RightArgumentTextBox.State.WaitForNotDisplayed(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        [Test]
+        public void Should_WaitForNotEnabled_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            RightArgumentTextBox.State.WaitForNotEnabled(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        private static void AssertIfWaitTimeIsCorrect(Stopwatch stopwatch)
```

Comment:
seems that *_CorrectAmountOfTime tests have similar actions and could be simplified with the following signature:

using IElementStateProvider = Aquality.Selenium.Core.Elements.Interfaces.IElementStateProvider;


        private static void AssertIfWaitTimeIsCorrect(IElementStateProvider stateProvider, Action<IElementStateProvider, TimeSpan> action)
        {
            var stopwatch = StartedStopwatch;
            action(stateProvider, FromSeconds);
            stopwatch.Stop();
            Assert.LessOrEqual(Math.Abs(stopwatch.Elapsed.Seconds - FromSeconds.Seconds), 1, "Wait Time is NOT correct");
        }
`

        [Test]
        public void Should_WaitForNotEnabled_CorrectAmountOfTime()
        {
            AssertIfWaitTimeIsCorrect(RightArgumentTextBox.State, (state, time) => state.WaitForNotEnabled(time));
        }

        [Test]
        public void Should_WaitForClickable_CorrectAmountOfTime()
        {
            AssertIfWaitTimeIsCorrect(EmptyButton.State, (state, time) =>
            {
                try
                {
                    EmptyButton.State.WaitForClickable(FromSeconds);
                }
                catch (WebDriverTimeoutException)
                {
                }
            });
        }

Reply:
done